components/canvas/raqote_backend.rs

```diff
@@ -557,6 +560,29 @@ impl GenericDrawTarget for raqote::DrawTarget {
             &DrawOptions::Raqote(draw_options),
         );
     }
+    fn fill_text(
+        &mut self,
+        text: String,
+        x: f32,
+        y: f32,
+        pattern: canvas_data::Pattern,
+        draw_options: &DrawOptions,
+    ) {
+        let font = SystemSource::new()
```

Comment:
Needs a TODO because this should eventually use the same infrastructure as layout (i.e. layout should be updated to use font-kit as well).

Reply:
Also, this is in the chrome process, right? It needs to not be in the content process because the content process shouldn't have access to the `SystemSource`. (It's too dangerous; on macOS for example it needs a connection to the font server, which has had a lot of security vulnerabilities.)

Reply:
Yes, the canvas code runs in the chrome process.

Reply:
So, I need to do:

1. add a TODO to `raqote_backend.rs`
2. move `font_kit` code from `raqote_backend.rs` to `canvas_data.rs` and pass the `SystemSource` things through` the parameter?

 

> Also, this is in the chrome process, right? It needs to not be in the content process because the content process shouldn't have access to the `SystemSource`.

I noticed there is a `FontCacheThread`.

Reply:
Let's keep the font_kit code in raqote_backend.rs for now until it's more clear how the proper font integration will work. The FontCacheThread is integrated with many other parts of the system, so converting that code to use font_kit will be part of resolving that TODO.

Reply:
@jdm 
If I understood well, I moved all font_kit code to `raqote_backend.rs`, including some functions: 
 `replace_whitespace` & `get_text_width`( They are parts of `text preparation algorithm`, in fact, I just moved the algorithm to `raqote_backend.rs`).